File: examples/android/app/src/main/java/org/aetherfeed/app/readerimport/FeedIdentity.kt
package org.aetherfeed.app.readerimport

import org.aetherfeed.app.domain.ModuleKind
import org.aetherfeed.app.news.OpmlOutline
import java.net.URI

private val podcastHosts = listOf(
    "libsyn.com",
    "megaphone.fm",
    "buzzsprout.com",
    "spreaker.com",
    "anchor.fm",
    "podbean.com",
    "simplecast.com",
    "transistor.fm",
    "captivate.fm",
    "omny.fm",
    "podcasts.apple.com",
    "pinecast.com",
    "art19.com",
)

fun stripFeedPrefix(url: String): String {
    val trimmed = url.trim()
    return if (trimmed.startsWith("feed/", ignoreCase = true)) trimmed.substring(5) else trimmed
}

fun normalizeFeedUrl(url: String): String {
    val raw = stripFeedPrefix(url)
    return runCatching {
        val withScheme = if ("://" in raw) raw else "https://$raw"
        val parsed = URI(withScheme)
        var host = (parsed.host ?: "").lowercase()
        if (host.startsWith("www.")) host = host.removePrefix("www.")
        val path = (parsed.path ?: "").trimEnd('/')
        val protocol = when (parsed.scheme?.lowercase()) {
            "http", "https" -> "https"
            else -> parsed.scheme ?: "https"
        }
        "$protocol://$host$path".lowercase()
    }.getOrElse { raw.lowercase().trimEnd('/') }
}

private fun hostOf(url: String): String? = runCatching {
    val raw = stripFeedPrefix(url)
    val withScheme = if ("://" in raw) raw else "https://$raw"
    var host = URI(withScheme).host?.lowercase() ?: return@runCatching null
    if (host.startsWith("www.")) host = host.removePrefix("www.")
    host
}.getOrNull()

fun classifyFeedKind(outline: OpmlOutline): ModuleKind {
    if (outline.type.equals("podcast", ignoreCase = true)) return ModuleKind.Podcast
    val folder = outline.folder.orEmpty()
    if (folder.contains("podcast", ignoreCase = true) || folder.contains("audio", ignoreCase = true)) {
        return ModuleKind.Podcast
    }
    if (outline.title.contains("podcast", ignoreCase = true)) return ModuleKind.Podcast
    val url = outline.xmlUrl.orEmpty()
    val host = hostOf(url)
    if (host != null && podcastHosts.any { host == it || host.endsWith(".$it") }) return ModuleKind.Podcast
    val path = runCatching {
        val raw = stripFeedPrefix(url)
        URI(if ("://" in raw) raw else "https://$raw").path.orEmpty()
    }.getOrDefault("")
    if (path.contains("/podcast", ignoreCase = true)) return ModuleKind.Podcast
    return ModuleKind.News
}

File: shared/typescript/feedIdentity.ts
/** Normalize subscription URLs and classify news vs podcast from an export. */

import type { ModuleKind } from "./models";
import type { OpmlOutline } from "./news";

const PODCAST_HOSTS = [
  "libsyn.com",
  "megaphone.fm",
  "buzzsprout.com",
  "spreaker.com",
  "anchor.fm",
  "podbean.com",
  "simplecast.com",
  "transistor.fm",
  "captivate.fm",
  "omny.fm",
  "podcasts.apple.com",
  "pinecast.com",
  "art19.com",
];

export function stripFeedPrefix(url: string): string {
  const trimmed = url.trim();
  return trimmed.toLowerCase().startsWith("feed/") ? trimmed.slice(5) : trimmed;
}

export function normalizeFeedUrl(url: string): string {
  const raw = stripFeedPrefix(url);
  try {
    const parsed = new URL(raw.includes("://") ? raw : `https://${raw}`);
    let host = parsed.hostname.toLowerCase();
    if (host.startsWith("www.")) host = host.slice(4);
    const path = parsed.pathname.replace(/\/+$/, "");
    const protocol = parsed.protocol === "http:" || parsed.protocol === "https:" ? "https:" : parsed.protocol;
    return `${protocol}//${host}${path}`.toLowerCase();
  } catch {
    return raw.toLowerCase().replace(/\/+$/, "");
  }
}

function hostOf(url: string): string | undefined {
  try {
    let host = new URL(stripFeedPrefix(url).includes("://") ? stripFeedPrefix(url) : `https://${url}`).hostname.toLowerCase();
    if (host.startsWith("www.")) host = host.slice(4);
    return host;
  } catch {
    return undefined;
  }
}

export function classifyFeedKind(outline: OpmlOutline): ModuleKind {
  const folder = outline.folder ?? "";
  const title = outline.title;
  if (/podcast|audio/i.test(folder) || /podcast/i.test(title)) return "podcast";
  const url = outline.xmlUrl ?? "";
  const host = hostOf(url);
  if (host && PODCAST_HOSTS.some((known) => host === known || host.endsWith(`.${known}`))) return "podcast";
  try {
    const path = new URL(stripFeedPrefix(url).includes("://") ? stripFeedPrefix(url) : `https://${url}`).pathname;
    if (/\/podcast/i.test(path)) return "podcast";
  } catch {
    /* keep news */
  }
  return "news";
}
